File: src/main/java/AoC2021_20.java
import static java.util.stream.Collectors.summarizingInt;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.IntSummaryStatistics;
import java.util.List;
import java.util.Set;

import org.apache.commons.math3.util.IntegerSequence.Range;

import com.github.pareronia.aoc.Grid.Cell;
import com.github.pareronia.aocd.Aocd;
import com.github.pareronia.aocd.Puzzle;

public class AoC2021_20 extends AoCBase {
    
    private static final char DARK = '.';
    private static final char LIGHT = '#';
    
    private final ImageEnhancement ie;
    
    private AoC2021_20(final List<String> input, final boolean debug) {
        super(debug);
        final String algorithm = input.get(0);
        assert algorithm.length() == 512;
        final Set<Cell> lights = new HashSet<>();
        final List<String> imageLines = input.subList(2, input.size());
        input.subList(2, input.size());
        for (int i = 0; i < imageLines.size(); i++) {
            final String line = imageLines.get(i);
            for (int j = 0; j < line.length(); j++) {
                if (line.charAt(j) == LIGHT) {
                    lights.add(Cell.at(i, j));
                }
            }
        }
        this.ie = new ImageEnhancement(algorithm, lights);
    }
    
    public static final AoC2021_20 create(final List<String> input) {
        return new AoC2021_20(input, false);
    }

    public static final AoC2021_20 createDebug(final List<String> input) {
        return new AoC2021_20(input, true);
    }
    
    private int solve(final int cycles) {
        for (int i = 0; i < cycles; i++) {
            ie.run();
        }
        return ie.getLitPixels();
    }

    @Override
    public Integer solvePart1() {
        return solve(2);
    }
    
    @Override
    public Integer solvePart2() {
        return solve(50);
    }

    public static void main(final String[] args) throws Exception {
        assert AoC2021_20.createDebug(TEST).solvePart1() == 35;
        assert AoC2021_20.create(TEST).solvePart2() == 3351;

        final Puzzle puzzle = Aocd.puzzle(2021, 20);
        puzzle.check(
            () -> lap("Part 1", () -> AoC2021_20.create(puzzle.getInputData()).solvePart1()),
            () -> lap("Part 2", () -> AoC2021_20.create(puzzle.getInputData()).solvePart2())
        );
    }

    private static final List<String> TEST = splitLines(
        "..#.#..#####.#.#.#.###.##.....###.##.#..###.####..#####..#....#..#..##..##" +
        "#..######.###...####..#..#####..##..#.#####...##.#.#..#.##..#.#......#.###" +
        ".######.###.####...#.##.##..#..#..#####.....#.#....###..#.##......#.....#." +
        ".#..#..##..#...##.######.####.####.#.#...#.......#..#.#.#...####.##.#....." +
        ".#..#...##.#.##..#...##.#.##..###.#......#.#.......#.#.#.####.###.##...#.." +
        "...####.#..#..#.##.#....##..#.####....##...##..#...#......#.#.......#....." +
        "..##..####..#...#.#.#...##..#.#..###..#####........#..####......#..#\r\n" +
        "\r\n" +
        "#..#.\r\n" +
        "#....\r\n" +
        "##..#\r\n" +
        "..#..\r\n" +
        "..###"
    );
    
    private static final class ImageEnhancement {
        private final String algorithm;
        private final boolean isFlicker;
        private Set<Cell> data;
        private boolean storeDark = false;
        
        public ImageEnhancement(final String algorithm, final Set<Cell> data) {
            this.algorithm = algorithm;
            this.isFlicker = algorithm.charAt(0) == LIGHT
                    && algorithm.charAt(algorithm.length() - 1) == DARK;
            this.data = data;
        }

        private Range getRowRange() {
            final IntSummaryStatistics summary = this.data.stream()
                    .map(Cell::getRow)
                    .collect(summarizingInt(Integer::intValue));
            return new Range(summary.getMin() - 2, summary.getMax() + 2, 1);
        }

        private Range getColRange() {
            final IntSummaryStatistics summary = this.data.stream()
                    .map(Cell::getCol)
                    .collect(summarizingInt(Integer::intValue));
            return new Range(summary.getMin() - 2, summary.getMax() + 2, 1);
        }
        
        @SuppressWarnings("unused")
        public List<String> print() {
            final List<String> lines = new ArrayList<>();
            final Range rowRange = this.getRowRange();
            final Range colRange = this.getColRange();
            for (final int row : rowRange) {
                final StringBuilder sb = new StringBuilder();
                for (final int col : colRange) {
                    sb.append(this.data.contains(Cell.at(row, col)) ? LIGHT : DARK);
                }
                lines.add(String.valueOf(sb.toString()));
            }
            return lines;
        }
        
        private String getSquareAround(final int row, final int col) {
            final char found;
            final char notFound;
            if (!this.isFlicker || !this.storeDark) {
                found = '1';
                notFound = '0';
            } else {
                found = '0';
                notFound = '1';
            }
            final StringBuilder sb = new StringBuilder();
            for (final int rr : List.of(-1, 0, 1)) {
                for (final int cc : List.of(-1, 0, 1)) {
                    sb.append(this.data.contains(Cell.at(row + rr, col + cc))
                                ? found : notFound);
                }
            }
            return sb.toString();
        }
        
        public void run() {
            final Set<Cell> data2 = new HashSet<>();
            final Range rowRange = this.getRowRange();
            final Range colRange = this.getColRange();
            for (final int row : rowRange) {
                for (final int col : colRange) {
                    final String square = getSquareAround(row, col);
                    final int idx = Integer.parseInt(square, 2);
                    assert idx >= 0 && idx < this.algorithm.length();
                    final boolean light = this.algorithm.charAt(idx) == LIGHT;
                    if (((!this.isFlicker || this.storeDark) && light)
                        || (this.isFlicker && !this.storeDark && !light)) {
                        data2.add(Cell.at(row, col));
                    }
                }
            }
            this.data = data2;
            this.storeDark = !this.storeDark;
        }
        
        public int getLitPixels() {
            return this.data.size();
        }
    }
}

File: src/main/python/AoC2021_20.py
#! /usr/bin/env python3
#
# Advent of Code 2021 Day 20
#

from prettyprinter import register_pretty, pretty_call
from aoc import my_aocd
import aocd


DARK = '.'
LIGHT = '#'


class ImageEnhancement:
    algorithm: str
    data: set[tuple[int, int]]
    store_dark: bool

    def __init__(self, algorithm: str, data: set[tuple[int, int]]):
        assert len(algorithm) == 512
        self.algorithm = algorithm
        self.data = data
        self.store_dark = False

    def _is_flicker(self) -> bool:
        return self.algorithm[0] == LIGHT and self.algorithm[-1] == DARK

    def _row_range(self):
        rows = {row for row, _ in self.data}
        return range(min(rows) - 2, max(rows) + 3)

    def _col_range(self):
        cols = {col for _, col in self.data}
        return range(min(cols) - 2, max(cols) + 3)

    def _print(self) -> None:
        if not __debug__:
            return
        row_range = self._row_range()
        col_range = self._col_range()
        lines = ["".join(LIGHT if (row, col) in self.data else DARK
                         for col in col_range)
                 for row in row_range]
        [print(line) for line in lines]

    def _get_square_around(self, row: int, col: int) -> str:
        found, not_found = ('1', '0') \
                if not self._is_flicker() or not self.store_dark \
                else ('0', '1')
        return "".join(found
                       if (row + dr, col + dc) in self.data
                       else not_found
                       for dr in [-1, 0, 1]
                       for dc in [-1, 0, 1])

    def run(self) -> None:
        data2 = set[tuple[int, int]]()
        row_range = self._row_range()
        col_range = self._col_range()
        for row in row_range:
            for col in col_range:
                square = self._get_square_around(row, col)
                idx = int(square, 2)
                assert 0 <= idx < len(self.algorithm)
                light = self.algorithm[idx] == LIGHT
                if ((not self._is_flicker() or self.store_dark) and light) or \
                   (self._is_flicker() and not self.store_dark and not light):
                    data2.add((row, col))
        self.data = data2
        self.store_dark = not self.store_dark

    def get_lit_pixels(self) -> int:
        return len(self.data)


@register_pretty(ImageEnhancement)
def _pretty(value, ctx):
    return pretty_call(
        ctx,
        ImageEnhancement,
        algorithm=value.algorithm,
        data=value.data
    )


def _parse(inputs: tuple[str]) -> int:
    blocks = my_aocd.to_blocks(inputs)
    algorithm = blocks[0][0]
    data = {(r, c)
            for r, line in enumerate(blocks[1])
            for c, ch in enumerate(line)
            if ch == LIGHT}
    return ImageEnhancement(algorithm, data)


def _solve(ie: ImageEnhancement, cycles: int) -> int:
    for i in range(cycles):
        ie.run()
    return ie.get_lit_pixels()


def part_1(inputs: tuple[str]) -> int:
    ie = _parse(inputs)
    return _solve(ie, 2)


def part_2(inputs: tuple[str]) -> int:
    ie = _parse(inputs)
    return _solve(ie, 50)


TEST = """\
..#.#..#####.#.#.#.###.##.....###.##.#..###.####..#####..#....#..#..##..###..######.###...####..#..#####..##..#.#####...##.#.#..#.##..#.#......#.###.######.###.####...#.##.##..#..#..#####.....#.#....###..#.##......#.....#..#..#..##..#...##.######.####.####.#.#...#.......#..#.#.#...####.##.#......#..#...##.#.##..#...##.#.##..###.#......#.#.......#.#.#.####.###.##...#.....####.#..#..#.##.#....##..#.####....##...##..#...#......#.#.......#.......##..####..#...#.#.#...##..#.#..###..#####........#..####......#..#

#..#.
#....
##..#
..#..
..###
""".splitlines()


def main() -> None:
    puzzle = aocd.models.Puzzle(2021, 20)
    my_aocd.print_header(puzzle.year, puzzle.day)

    assert part_1(TEST) == 35
    assert part_2(TEST) == 3351

    inputs = my_aocd.get_input(puzzle.year, puzzle.day, 102)
    result1 = part_1(inputs)
    print(f"Part 1: {result1}")
    result2 = part_2(inputs)
    print(f"Part 2: {result2}")
    my_aocd.check_results(puzzle, result1, result2)


if __name__ == '__main__':
    main()
